File: worker/src/worker/hasher.py
"""Streaming SHA-256 over a binary file-like object.

The next.js upload caps files at 10 MB, so reading into memory is fine in
practice — but streaming avoids ever holding both the file *and* a hashed
copy if Docling needs to materialise the file on disk anyway.
"""

from __future__ import annotations

import hashlib


def hash_stream(stream) -> tuple[str, bytes]:
    """Consume `stream`, return (hex_sha256, raw_bytes).

    `stream` is consumed in 1 MiB chunks until EOF. Returns the full
    payload alongside the digest so callers can hand the bytes to a
    tempfile or directly to Docling without re-reading the object.
    """

    hasher = hashlib.sha256()
    chunks: list[bytes] = []
    for chunk in iter(lambda: stream.read(1024 * 1024), b""):
        hasher.update(chunk)
        chunks.append(chunk)
    return hasher.hexdigest(), b"".join(chunks)

File: worker-rs/src/hasher.rs
//! Streaming SHA-256 over a `Bytes`-like buffer.
//!
//! `docling-rs` and the S3 client both hand us the source bytes as a
//! single in-memory `Vec<u8>` or `Bytes`; we still use the streaming API
//! because (a) it scales to large PDFs without an extra copy and (b) it
//! makes the hashing step composable if we ever swap to a streaming fetch.

use sha2::{Digest, Sha256};

/// Compute the SHA-256 hex digest of `bytes`. Returns `(hex, len)`.
pub fn hash_bytes(bytes: &[u8]) -> (String, usize) {
    let mut hasher = Sha256::new();
    hasher.update(bytes);
    let digest = hasher.finalize();
    (hex::encode(digest), bytes.len())
}